WeRadiate-ThermoSense.ino

```diff
@@ -405,6 +464,214 @@ void loop(void)
+        || to support plug/unplug and the sw interface is not
+        || really hot-pluggable.
+        */
+        
+        /* set D11 high so V_OUT2 is going to be high for onewire sensor */        
+        pinMode(D11, OUTPUT);
+        digitalWrite(D11, HIGH);
+        
+        bool fCompostTemp = checkCompostSensorPresent();
+      
+        if (fCompostTemp)
+                {
+                sensor_CompostTemp.requestTemperatures();
+                float compostTempC = sensor_CompostTemp.getTempCByIndex(0);
+                Serial.print("Compost temperature: "); Serial.print(compostTempC); Serial.println(" C");
+                // transmit the measurement
+                b.putT(compostTempC);
+                flag |= FlagsSensor3::FlagWater;
+                }
+        else if (fHasCompostTemp)
+                {
+                gCatena.SafePrintf("No compost temperature\n");
+                }
+             
+        /* set D11 low to turn off after measuring */        
+//        digitalWrite(D11, LOW);
+        pinMode(D11, INPUT);
+      
+        *pFlag = uint8_t(flag);
+        if (savedLed != LedPattern::Joining)
+                {
+                gLed.Set(LedPattern::Sending);
+                }
+        else
+                {
+                gLed.Set(LedPattern::Joining);
+                }
+        
+        gLoRaWAN.SendBuffer(b.getbase(), b.getn(), sendBufferDoneCb, NULL);
+}
+
+void
+sendBufferDoneCb(
+        void *pContext,
+        bool fStatus
+        )
+        {
+        osjobcb_t pFn;
+
+        gLed.Set(LedPattern::Settling);
+        if (! fStatus)
+                {
+                gCatena.SafePrintf("send buffer failed\n");
+                pFn = txFailedDoneCb;
+                }
+        else
+                {
+                pFn = settleDoneCb;
+                }
+        os_setTimedCallback(
+                &sensorJob,
+                os_getTime()+sec2osticks(CATCFG_T_SETTLE),
+                pFn
+                );
+        }
+
+void
+txFailedDoneCb(
+        osjob_t *pSendJob
+        )
+        {
+        gCatena.SafePrintf("not provisioned, idling\n");
+        gLoRaWAN.Shutdown();
+        gLed.Set(LedPattern::NotProvisioned);
+        }
+
+void settleDoneCb(
+        osjob_t *pSendJob
+        )
+        {
+        // if connected to USB, don't sleep
+        // ditto if we're monitoring pulses.
+	// XXX: do we need this?
+#ifdef USBCON
+        if (fUsbPower || fHasPower1)
+#else
+        if (fHasPower1)
+#endif
+        {
+        gLed.Set(LedPattern::Sleeping);
+        os_setTimedCallback(
+                &sensorJob,
+                os_getTime() + sec2osticks(CATCFG_T_INTERVAL),
+                sleepDoneCb
+                );
+        return;
+        }
+
+        /* ok... now it's time for a deep sleep */
+        gLed.Set(LedPattern::Off);
+        Serial.end();
+        Wire.end();
+        SPI.end();
+	if (fFlash)
```

Comment:
Because it's so important, I would put lines 640 through 642 in a separate subroutine (`prepareToSleep()`), and in that routine I would also make sure D11 was low, that the BME280 and SI1133 are in standby, and that the flash is powered down.

Reply:
void prepareToSleep(void)
        {
        Serial.end();
        Wire.end();
        SPI.end();
        if (digitalRead(D11) == HIGH)
                pinMode(D11, INPUT);
        else if (fFlash)
                gSPI2.end();
        }

Reply:
is this okay? and how do I make sure that BME280 and Si1133 are in standby?

Reply:
You need to send the standby commands (there are separate commands for each. We do that currently in the startup code).